mlflow/crewai/__init__.py

```diff
@@ -0,0 +1,169 @@
+):
+    """
+    Enables (or disables) and configures autologging from CrewAI to MLflow.
+    Only synchronous calls are supported. Asynchnorous APIs and streaming are not recorded.
+
+    Args:
+        log_traces: If ``True``, traces are logged for CrewAI agents.
+            If ``False``, no traces are collected during inference. Default to ``True``.
+        disable: If ``True``, disables the CrewAI autologging. Default to ``False``.
+        silent: If ``True``, suppress all event logs and warnings from MLflow during CrewAI
+            autologging. If ``False``, show all events and warnings.
+    """
+    # TODO: handle asynchronous tasks and crew executions
+    # TODO: interface of tool in CrewAI is changing drastically. Add patching once it's stabilized
+    import crewai
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Crew,
+        "kickoff",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Crew,
+        "kickoff_for_each",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Crew,
+        "train",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Agent,
+        "execute_task",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Task,
+        "execute_sync",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.LLM,
+        "call",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Flow,
+        "kickoff",
+        patched_class_call,
+    )
+    try:
+        # knowledge and memory are not available before 0.83.0
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.ShortTermMemory,
+            "save",
+            patched_class_call,
+        )
+
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.ShortTermMemory,
+            "search",
+            patched_class_call,
+        )
+
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.ShortTermMemory,
+            "reset",
+            patched_class_call,
+        )
+
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.LongTermMemory,
+            "save",
+            patched_class_call,
+        )
+
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.LongTermMemory,
+            "search",
+            patched_class_call,
+        )
+
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.LongTermMemory,
+            "reset",
+            patched_class_call,
+        )
+
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.UserMemory,
+            "save",
+            patched_class_call,
+        )
+
+        safe_patch(
+            FLAVOR_NAME,
+            crewai.memory.UserMemory,
+            "search",
+            patched_class_call,
+        )
+
+        safe_patch(
```

Comment:
Is there any limit (either theoretical or practical) for number of entities saved to `EntityMemory` in one task? If it can go sth like 100>, we probably want to either not patching or organize them into a parent span, otherwise the span UI becomes hart to navigate.

Reply:
There are only 4 types of memories and agent just call several methods for each, I don't think it will go that many.

Reply:
You're right, but is there any better way to make this more readable? We shouldn't change the span structure, but there is nothing bad in adding a parent span for a real function call or changing span names.

![Screenshot 2024-12-02 at 9 38 03](https://github.com/user-attachments/assets/c287fe8c-b8c6-4403-b1e0-229d6604a7f7)